File: src/strategies/listing-pump.ts
/**
 * Listing Pump Strategy (ported from v1 + enhanced)
 *
 * Detects new exchange listings across Coinbase, Binance, and Kraken.
 * New listings historically pump 20-100% in the first 48h.
 *
 * Enhancements over v1:
 *  - Multi-exchange: catches Binance listings (not just Coinbase)
 *  - Pre-listing signal: monitors CryptoPanic for "listing" news before official
 *  - Checks if token already listed elsewhere (less pumpy if so)
 *  - Estimates first-mover opportunity via time-since-announcement
 */

import type { TradeSignal } from '../types.js';
import { randomUUID } from 'crypto';

export interface ListingAnnouncement {
  symbol: string;
  exchange: 'coinbase' | 'binance' | 'kraken' | 'bybit';
  productId: string;
  announcedAt: number;
  tradingStartsAt?: number;
  isNewToMajorExchanges: boolean; // false if already on Binance when listing Coinbase
}

const seenListings = new Set<string>();
const LISTING_EXPIRY_MS = 48 * 3_600_000;

export function onListingAnnouncement(listing: ListingAnnouncement, currentPrice: number): TradeSignal | null {
  const key = `${listing.exchange}:${listing.symbol}`;
  if (seenListings.has(key)) return null;
  seenListings.add(key);

  const ageMs = Date.now() - listing.announcedAt;
  if (ageMs > LISTING_EXPIRY_MS) return null; // too stale

  // Base score: Coinbase listings have strongest pump history
  let baseScore = 55;
  switch (listing.exchange) {
    case 'coinbase': baseScore = 75; break;
    case 'binance':  baseScore = 72; break;
    case 'kraken':   baseScore = 60; break;
    case 'bybit':    baseScore = 58; break;
  }

  // First mover bonus: within 30 min of announcement = max bonus
  const freshnessBonus = Math.max(0, 15 - Math.floor(ageMs / 120_000));

  // Penalty if already on major exchanges (less retail discovery excitement)
  const alreadyListedPenalty = listing.isNewToMajorExchanges ? 0 : -15;

  const score = Math.min(95, baseScore + freshnessBonus + alreadyListedPenalty);

  return {
    id: randomUUID(),
    symbol: listing.symbol,
    productId: listing.productId,
    strategy: 'listing_pump',
    side: 'long',
    tier: 'swing',
    score,
    confidence: score > 80 ? 'high' : score > 65 ? 'medium' : 'low',
    sources: ['listing_detector'],
    reasoning: `${listing.symbol} new ${listing.exchange} listing announced ${Math.floor(ageMs / 60_000)}m ago${listing.isNewToMajorExchanges ? ' (first major exchange)' : ''}`,
    entryPrice: currentPrice,
    stopPrice: currentPrice * 0.88, // wider stop — listing pumps can be volatile
    suggestedSizeUsd: 120,
    expiresAt: Date.now() + LISTING_EXPIRY_MS,
    createdAt: Date.now(),
  };
}

File: src/strategies/listing_pump.py
"""Listing Pump Strategy."""

import threading
import time
import uuid
from dataclasses import dataclass
from typing import Optional

from src.types import TradeSignal


@dataclass
class ListingAnnouncement:
    symbol: str
    exchange: str
    product_id: str
    announced_at: float
    is_new_to_major_exchanges: bool
    trading_starts_at: Optional[float] = None


_seen_listings: set[str] = set()
_listing_timestamps: dict[str, float] = {}
_lock = threading.Lock()
_LISTING_EXPIRY_MS = 6 * 3_600_000  # 6 hours — listing pumps exhaust quickly
_last_cleanup: float = 0

STRATEGY_META = {
    "strategies": [
        {"id": "listing_announcement", "function": "on_listing_announcement",
         "description": "Detects new exchange listings and rides the initial pump",
         "tier": "swing"},
    ],
    "signal_sources": ["price_action"],
}


def _cleanup_expired_listings(now: float) -> None:
    global _last_cleanup
    if now - _last_cleanup < 3_600_000:
        return
    _last_cleanup = now
    expired = [k for k, ts in _listing_timestamps.items() if now - ts > _LISTING_EXPIRY_MS]
    for k in expired:
        _seen_listings.discard(k)
        _listing_timestamps.pop(k, None)


def on_listing_announcement(listing: ListingAnnouncement, current_price: float) -> Optional[TradeSignal]:
    now = time.time() * 1000
    key = f"{listing.exchange}:{listing.symbol}"

    with _lock:
        _cleanup_expired_listings(now)
        if key in _seen_listings:
            return None
        _seen_listings.add(key)
        _listing_timestamps[key] = now

    age_ms = now - listing.announced_at
    if age_ms > _LISTING_EXPIRY_MS:
        return None

    base_score = {"coinbase": 75, "binance": 72, "kraken": 60, "bybit": 58}.get(listing.exchange, 55)
    # Decay faster: penalize signals older than 15 minutes instead of rewarding late entries
    age_minutes = age_ms / 60_000
    if age_minutes <= 15:
        freshness_bonus = max(0, 15 - int(age_minutes))
    else:
        freshness_bonus = -min(10, int((age_minutes - 15) / 5))
    already_listed_penalty = 0 if listing.is_new_to_major_exchanges else -15
    score = min(95, base_score + freshness_bonus + already_listed_penalty)

    return TradeSignal(
        id=str(uuid.uuid4()), symbol=listing.symbol, product_id=listing.product_id,
        strategy="listing_pump", side="long", tier="swing", score=score,
        confidence="high" if score > 80 else ("medium" if score > 65 else "low"),
        sources=["listing_detector"],
        reasoning=f"{listing.symbol} new {listing.exchange} listing {int(age_ms/60_000)}m ago"
                  + (" (first major exchange)" if listing.is_new_to_major_exchanges else ""),
        entry_price=current_price, stop_price=current_price * 0.85,
        target_price=current_price * 1.20,  # R:R fix: 20% target vs 15% stop = 1.33:1
        suggested_size_usd=120,
        expires_at=now + _LISTING_EXPIRY_MS, created_at=now,
    )
